File: dotnet/src/support/Colors/Color.cs
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Globalization;
using System.Text.RegularExpressions;

namespace OpenQA.Selenium.Support.Colors
{
    /// <summary>
    /// Color conversion support class.
    /// </summary>
    public class Color
    {
        public int Red { get; }
        public int Green { get; }
        public int Blue { get; }
        public double Alpha { get; set; }

        public string Rgb => $"rgb({Red}, {Green}, {Blue})";
        public string Rgba => $"rgba({Red}, {Green}, {Blue}, {Alpha.ToString(CultureInfo.InvariantCulture)})";
        public string Hex => $"#{Red:X2}{Green:X2}{Blue:X2}".ToLower();
        public System.Drawing.Color GetColor => System.Drawing.Color.FromArgb((int)(Alpha * 255), Red, Green, Blue);

        private static Dictionary<string, Color> Colors = new()
    {
        { "TRANSPARENT", new Color(0, 0, 0, 0d) },
        { "ALICEBLUE", new Color(240, 248, 255, 1d) },
        { "ANTIQUEWHITE", new Color(250, 235, 215, 1d) },
        { "AQUA", new Color(0, 255, 255, 1d) },
        { "AQUAMARINE", new Color(127, 255, 212, 1d) },
        { "AZURE", new Color(240, 255, 255, 1d) },
        { "BEIGE", new Color(245, 245, 220, 1d) },
        { "BISQUE", new Color(255, 228, 196, 1d) },
        { "BLACK", new Color(0, 0, 0, 1d) },
        { "BLANCHEDALMOND", new Color(255, 235, 205, 1d) },
        { "BLUE", new Color(0, 0, 255, 1d) },
        { "BLUEVIOLET", new Color(138, 43, 226, 1d) },
        { "BROWN", new Color(165, 42, 42, 1d) },
        { "BURLYWOOD", new Color(222, 184, 135, 1d) },
        { "CADETBLUE", new Color(95, 158, 160, 1d) },
        { "CHARTREUSE", new Color(127, 255, 0, 1d) },
        { "CHOCOLATE", new Color(210, 105, 30, 1d) },
        { "CORAL", new Color(255, 127, 80, 1d) },
        { "CORNFLOWERBLUE", new Color(100, 149, 237, 1d) },
        { "CORNSILK", new Color(255, 248, 220, 1d) },
        { "CRIMSON", new Color(220, 20, 60, 1d) },
        { "CYAN", new Color(0, 255, 255, 1d) },
        { "DARKBLUE", new Color(0, 0, 139, 1d) },
        { "DARKCYAN", new Color(0, 139, 139, 1d) },
        { "DARKGOLDENROD", new Color(184, 134, 11, 1d) },
        { "DARKGRAY", new Color(169, 169, 169, 1d) },
        { "DARKGREEN", new Color(0, 100, 0, 1d) },
        { "DARKGREY", new Color(169, 169, 169, 1d) },
        { "DARKKHAKI", new Color(189, 183, 107, 1d) },
        { "DARKMAGENTA", new Color(139, 0, 139, 1d) },
        { "DARKOLIVEGREEN", new Color(85, 107, 47, 1d) },
        { "DARKORANGE", new Color(255, 140, 0, 1d) },
        { "DARKORCHID", new Color(153, 50, 204, 1d) },
        { "DARKRED", new Color(139, 0, 0, 1d) },
        { "DARKSALMON", new Color(233, 150, 122, 1d) },
        { "DARKSEAGREEN", new Color(143, 188, 143, 1d) },
        { "DARKSLATEBLUE", new Color(72, 61, 139, 1d) },
        { "DARKSLATEGRAY", new Color(47, 79, 79, 1d) },
        { "DARKSLATEGREY", new Color(47, 79, 79, 1d) },
        { "DARKTURQUOISE", new Color(0, 206, 209, 1d) },
        { "DARKVIOLET", new Color(148, 0, 211, 1d) },
        { "DEEPPINK", new Color(255, 20, 147, 1d) },
        { "DEEPSKYBLUE", new Color(0, 191, 255, 1d) },
        { "DIMGRAY", new Color(105, 105, 105, 1d) },
        { "DIMGREY", new Color(105, 105, 105, 1d) },
        { "DODGERBLUE", new Color(30, 144, 255, 1d) },
        { "FIREBRICK", new Color(178, 34, 34, 1d) },
        { "FLORALWHITE", new Color(255, 250, 240, 1d) },
        { "FORESTGREEN", new Color(34, 139, 34, 1d) },
        { "FUCHSIA", new Color(255, 0, 255, 1d) },
        { "GAINSBORO", new Color(220, 220, 220, 1d) },
        { "GHOSTWHITE", new Color(248, 248, 255, 1d) },
        { "GOLD", new Color(255, 215, 0, 1d) },
        { "GOLDENROD", new Color(218, 165, 32, 1d) },
        { "GRAY", new Color(128, 128, 128, 1d) },
        { "GREY", new Color(128, 128, 128, 1d) },
        { "GREEN", new Color(0, 128, 0, 1d) },
        { "GREENYELLOW", new Color(173, 255, 47, 1d) },
        { "HONEYDEW", new Color(240, 255, 240, 1d) },
        { "HOTPINK", new Color(255, 105, 180, 1d) },
        { "INDIANRED", new Color(205, 92, 92, 1d) },
        { "INDIGO", new Color(75, 0, 130, 1d) },
        { "IVORY", new Color(255, 255, 240, 1d) },
        { "KHAKI", new Color(240, 230, 140, 1d) },
        { "LAVENDER", new Color(230, 230, 250, 1d) },
        { "LAVENDERBLUSH", new Color(255, 240, 245, 1d) },
        { "LAWNGREEN", new Color(124, 252, 0, 1d) },
        { "LEMONCHIFFON", new Color(255, 250, 205, 1d) },
        { "LIGHTBLUE", new Color(173, 216, 230, 1d) },
        { "LIGHTCORAL", new Color(240, 128, 128, 1d) },
        { "LIGHTCYAN", new Color(224, 255, 255, 1d) },
        { "LIGHTGOLDENRODYELLOW", new Color(250, 250, 210, 1d) },
        { "LIGHTGRAY", new Color(211, 211, 211, 1d) },
        { "LIGHTGREEN", new Color(144, 238, 144, 1d) },
        { "LIGHTGREY", new Color(211, 211, 211, 1d) },
        { "LIGHTPINK", new Color(255, 182, 193, 1d) },
        { "LIGHTSALMON", new Color(255, 160, 122, 1d) },
        { "LIGHTSEAGREEN", new Color(32, 178, 170, 1d) },
        { "LIGHTSKYBLUE", new Color(135, 206, 250, 1d) },
        { "LIGHTSLATEGRAY", new Color(119, 136, 153, 1d) },
        { "LIGHTSLATEGREY", new Color(119, 136, 153, 1d) },
        { "LIGHTSTEELBLUE", new Color(176, 196, 222, 1d) },
        { "LIGHTYELLOW", new Color(255, 255, 224, 1d) },
        { "LIME", new Color(0, 255, 0, 1d) },
        { "LIMEGREEN", new Color(50, 205, 50, 1d) },
        { "LINEN", new Color(250, 240, 230, 1d) },
        { "MAGENTA", new Color(255, 0, 255, 1d) },
        { "MAROON", new Color(128, 0, 0, 1d) },
        { "MEDIUMAQUAMARINE", new Color(102, 205, 170, 1d) },
        { "MEDIUMBLUE", new Color(0, 0, 205, 1d) },
        { "MEDIUMORCHID", new Color(186, 85, 211, 1d) },
        { "MEDIUMPURPLE", new Color(147, 112, 219, 1d) },
        { "MEDIUMSEAGREEN", new Color(60, 179, 113, 1d) },
        { "MEDIUMSLATEBLUE", new Color(123, 104, 238, 1d) },
        { "MEDIUMSPRINGGREEN", new Color(0, 250, 154, 1d) },
        { "MEDIUMTURQUOISE", new Color(72, 209, 204, 1d) },
        { "MEDIUMVIOLETRED", new Color(199, 21, 133, 1d) },
        { "MIDNIGHTBLUE", new Color(25, 25, 112, 1d) },
        { "MINTCREAM", new Color(245, 255, 250, 1d) },
        { "MISTYROSE", new Color(255, 228, 225, 1d) },
        { "MOCCASIN", new Color(255, 228, 181, 1d) },
        { "NAVAJOWHITE", new Color(255, 222, 173, 1d) },
        { "NAVY", new Color(0, 0, 128, 1d) },
        { "OLDLACE", new Color(253, 245, 230, 1d) },
        { "OLIVE", new Color(128, 128, 0, 1d) },
        { "OLIVEDRAB", new Color(107, 142, 35, 1d) },
        { "ORANGE", new Color(255, 165, 0, 1d) },
        { "ORANGERED", new Color(255, 69, 0, 1d) },
        { "ORCHID", new Color(218, 112, 214, 1d) },
        { "PALEGOLDENROD", new Color(238, 232, 170, 1d) },
        { "PALEGREEN", new Color(152, 251, 152, 1d) },
        { "PALETURQUOISE", new Color(175, 238, 238, 1d) },
        { "PALEVIOLETRED", new Color(219, 112, 147, 1d) },
        { "PAPAYAWHIP", new Color(255, 239, 213, 1d) },
        { "PEACHPUFF", new Color(255, 218, 185, 1d) },
        { "PERU", new Color(205, 133, 63, 1d) },
        { "PINK", new Color(255, 192, 203, 1d) },
        { "PLUM", new Color(221, 160, 221, 1d) },
        { "POWDERBLUE", new Color(176, 224, 230, 1d) },
        { "PURPLE", new Color(128, 0, 128, 1d) },
        { "REBECCAPURPLE", new Color(102, 51, 153, 1d) },
        { "RED", new Color(255, 0, 0, 1d) },
        { "ROSYBROWN", new Color(188, 143, 143, 1d) },
        { "ROYALBLUE", new Color(65, 105, 225, 1d) },
        { "SADDLEBROWN", new Color(139, 69, 19, 1d) },
        { "SALMON", new Color(250, 128, 114, 1d) },
        { "SANDYBROWN", new Color(244, 164, 96, 1d) },
        { "SEAGREEN", new Color(46, 139, 87, 1d) },
        { "SEASHELL", new Color(255, 245, 238, 1d) },
        { "SIENNA", new Color(160, 82, 45, 1d) },
        { "SILVER", new Color(192, 192, 192, 1d) },
        { "SKYBLUE", new Color(135, 206, 235, 1d) },
        { "SLATEBLUE", new Color(106, 90, 205, 1d) },
        { "SLATEGRAY", new Color(112, 128, 144, 1d) },
        { "SLATEGREY", new Color(112, 128, 144, 1d) },
        { "SNOW", new Color(255, 250, 250, 1d) },
        { "SPRINGGREEN", new Color(0, 255, 127, 1d) },
        { "STEELBLUE", new Color(70, 130, 180, 1d) },
        { "TAN", new Color(210, 180, 140, 1d) },
        { "TEAL", new Color(0, 128, 128, 1d) },
        { "THISTLE", new Color(216, 191, 216, 1d) },
        { "TOMATO", new Color(255, 99, 71, 1d) },
        { "TURQUOISE", new Color(64, 224, 208, 1d) },
        { "VIOLET", new Color(238, 130, 238, 1d) },
        { "WHEAT", new Color(245, 222, 179, 1d) },
        { "WHITE", new Color(255, 255, 255, 1d) },
        { "WHITESMOKE", new Color(245, 245, 245, 1d) },
        { "YELLOW", new Color(255, 255, 0, 1d) },
        { "YELLOWGREEN", new Color(154, 205, 50, 1d) }
    };

        public Color(int red, int green, int blue, double alpha)
        {
            Red = red;
            Green = green;
            Blue = blue;
            Alpha = alpha;
        }

        public override string ToString()
        {
            return "Color: " + Rgba;
        }

        public override bool Equals(object? obj)
        {
            if (obj is not Color)
            {
                return false;
            }
            return Rgba.Equals(((Color)obj).Rgba);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Red, Green, Blue, Alpha);
        }

        /// <summary>
        /// Guesses what format the input color is in.
        /// </summary>
        /// <param name="value">A string representation of the color.</param>
        /// <returns>An instance of the <see cref="Color"/> class.</returns>
        /// <exception cref="ArgumentException"></exception>
        public static Color FromString(string value)
        {
            List<Converter> converters = new()
        {
            new RgbConverter(),
            new RgbPctConverter(),
            new RgbaConverter(),
            new RgbaPctConverter(),
            new NamedColorConverter(),
            new HexConverter(),
            new Hex3Converter(),
            new HslConverter(),
            new HslaConverter()
        };

            foreach (var converter in converters)
            {
                Color? color = converter.GetColor(value);
                if (color != null)
                {
                    return color;
                }
            }
            throw new ArgumentException($"Did not know how to convert {value} into color");
        }

        private abstract class Converter
        {
            public virtual Color? GetColor(string value)
            {
                Match match = GetRegex().Match(value);
                if (match.Success)
                {
                    double alpha = 1.0;
                    if (match.Groups.Count == 5)
                    {
                        alpha = double.Parse(match.Groups[4].Value, CultureInfo.InvariantCulture);
                    }
                    return CreateColor(match, alpha);
                }
                return null;
            }

            protected virtual Color CreateColor(Match match, double alpha)
            {
                return new Color(
                    FromMatchGroup(match, 1),
                    FromMatchGroup(match, 2),
                    FromMatchGroup(match, 3),
                    alpha);
            }

            protected virtual int FromMatchGroup(Match match, int index)
            {
                return int.Parse(match.Groups[index].Value);
            }

            protected abstract Regex GetRegex();
        }

        private class RgbConverter : Converter
        {
            protected override Regex GetRegex()
            {
                Regex rgbRegex = new(@"^\s*rgb\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*\)\s*$");
                return rgbRegex;
            }
        }

        private class RgbPctConverter : Converter
        {
            protected override Regex GetRegex()
            {
                Regex rgbPctRegex = new(@"^\s*rgb\(\s*(\d{1,3}|\d{1,2}\.\d+)%\s*,\s*(\d{1,3}|\d{1,2}\.\d+)%\s*,\s*(\d{1,3}|\d{1,2}\.\d+)%\s*\)\s*$");
                return rgbPctRegex;
            }

            protected override int FromMatchGroup(Match match, int index)
            {
                double n = double.Parse(match.Groups[index].Value) / 100 * 255;
                return (int)n;
            }
        }

        private class RgbaConverter : RgbConverter
        {
            protected override Regex GetRegex()
            {
                Regex rgbaRegex = new(@"^\s*rgba\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(0|1|0\.\d+)\s*\)\s*$");
                return rgbaRegex;
            }
        }

        private class RgbaPctConverter : RgbPctConverter
        {
            protected override Regex GetRegex()
            {
                Regex rgbaPctRegex = new(@"^\s*rgba\(\s*(\d{1,3}|\d{1,2}\.\d+)%\s*,\s*(\d{1,3}|\d{1,2}\.\d+)%\s*,\s*(\d{1,3}|\d{1,2}\.\d+)%\s*,\s*(0|1|0\.\d+)\s*\)\s*$");
                return rgbaPctRegex;
            }
        }

        private class HexConverter : Converter
        {
            protected override Regex GetRegex()
            {
                Regex hexRegex = new(@"#([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})");
                return hexRegex;
            }

            protected override int FromMatchGroup(Match match, int index)
            {
                return int.Parse(match.Groups[index].Value, NumberStyles.HexNumber);
            }
        }

        private class Hex3Converter : Converter
        {
            protected override Regex GetRegex()
            {
                Regex hex3Regex = new(@"#([A-Fa-f0-9])([A-Fa-f0-9])([A-Fa-f0-9])");
                return hex3Regex;
            }

            protected override int FromMatchGroup(Match match, int index)
            {
                return int.Parse(match.Groups[index].Value + match.Groups[index].Value, NumberStyles.HexNumber);
            }
        }

        private class HslConverter : Converter
        {
            protected override Regex GetRegex()
            {
                Regex hslRegex = new(@"^\s*hsl\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*\)\s*$");
                return hslRegex;
            }

            protected override Color CreateColor(Match match, double alpha)
            {
                double h = double.Parse(match.Groups[1].Value) / 360;
                double s = double.Parse(match.Groups[2].Value) / 100;
                double l = double.Parse(match.Groups[3].Value) / 100;
                double r, g, b;

                if (s == 0)
                {
                    r = l;
                    g = r;
                    b = r;
                }
                else
                {
                    double luminocity2 = (l < 0.5) ? l * (1 + s) : l + s - l * s;
                    double luminocity1 = 2 * l - luminocity2;
                    r = HueToRgb(luminocity1, luminocity2, h + 1.0 / 3.0);
                    g = HueToRgb(luminocity1, luminocity2, h);
                    b = HueToRgb(luminocity1, luminocity2, h - 1.0 / 3.0);
                }

                return new Color(
                    (int)Math.Round(r * 255),
                    (int)Math.Round(g * 255),
                    (int)Math.Round(b * 255),
                    alpha);
            }

            private double HueToRgb(double luminocity1, double luminocity2, double hue)
            {
                if (hue < 0.0)
                {
                    hue += 1;
                }
                if (hue > 1.0)
                {
                    hue -= 1;
                }
                if (hue < 1.0 / 6.0)
                {
                    return (luminocity1 + (luminocity2 - luminocity1) * 6.0 * hue);
                }
                if (hue < 1.0 / 2.0)
                {
                    return luminocity2;
                }
                if (hue < 2.0 / 3.0)
                {
                    return (luminocity1 + (luminocity2 - luminocity1) * ((2.0 / 3.0) - hue) * 6.0);
                }
                return luminocity1;
            }
        }

        private class HslaConverter : HslConverter
        {
            protected override Regex GetRegex()
            {
                Regex hslaRegex = new(@"^\s*hsla\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*,\s*(0|1|0\.\d+)\s*\)\s*$");
                return hslaRegex;
            }
        }

        private class NamedColorConverter : Converter
        {
            public override Color? GetColor(string value)
            {
                var valueUpper = value.ToUpper();
                if (Colors.ContainsKey(valueUpper))
                {
                    return Colors[valueUpper];
                }
                return null;
            }

            protected override Regex GetRegex()
            {
                throw new NotSupportedException("GetRegex is not supported");
            }
        }
    }
}

File: javascript/atoms-ts/src/color.ts
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

/**
 * Utilities for color parsing and standardization.
 * Converts various color formats (hex, rgb, rgba, named colors) to a standard rgba format.
 */

/**
 * RGBA color represented as a tuple of [red, green, blue, alpha]
 * where r, g, b are integers in [0, 255] and a is a float in [0, 1].
 */
type RgbaColor = [number, number, number, number];

/**
 * Standard CSS color names mapped to hex values.
 * This is a subset of the most commonly used CSS named colors.
 * For a complete list, use a dedicated color names package.
 */
const COLOR_NAMES: Record<string, string> = {
    aliceblue: '#f0f8ff',
    antiquewhite: '#faebd7',
    aqua: '#00ffff',
    aquamarine: '#7fffd4',
    azure: '#f0ffff',
    beige: '#f5f5dc',
    bisque: '#ffe4c4',
    black: '#000000',
    blanchedalmond: '#ffebcd',
    blue: '#0000ff',
    blueviolet: '#8a2be2',
    brown: '#a52a2a',
    burlywood: '#deb887',
    cadetblue: '#5f9ea0',
    chartreuse: '#7fff00',
    chocolate: '#d2691e',
    coral: '#ff7f50',
    cornflowerblue: '#6495ed',
    cornsilk: '#fff8dc',
    crimson: '#dc143c',
    cyan: '#00ffff',
    darkblue: '#00008b',
    darkcyan: '#008b8b',
    darkgoldenrod: '#b8860b',
    darkgray: '#a9a9a9',
    darkgrey: '#a9a9a9',
    darkgreen: '#006400',
    darkkhaki: '#bdb76b',
    darkmagenta: '#8b008b',
    darkolivegreen: '#556b2f',
    darkorange: '#ff8c00',
    darkorchid: '#9932cc',
    darkred: '#8b0000',
    darksalmon: '#e9967a',
    darkseagreen: '#8fbc8f',
    darkslateblue: '#483d8b',
    darkslategray: '#2f4f4f',
    darkslategrey: '#2f4f4f',
    darkturquoise: '#00ced1',
    darkviolet: '#9400d3',
    deeppink: '#ff1493',
    deepskyblue: '#00bfff',
    dimgray: '#696969',
    dimgrey: '#696969',
    dodgerblue: '#1e90ff',
    firebrick: '#b22222',
    floralwhite: '#fffaf0',
    forestgreen: '#228b22',
    fuchsia: '#ff00ff',
    gainsboro: '#dcdcdc',
    ghostwhite: '#f8f8ff',
    gold: '#ffd700',
    goldenrod: '#daa520',
    gray: '#808080',
    grey: '#808080',
    green: '#008000',
    greenyellow: '#adff2f',
    honeydew: '#f0fff0',
    hotpink: '#ff69b4',
    indianred: '#cd5c5c',
    indigo: '#4b0082',
    ivory: '#fffff0',
    khaki: '#f0e68c',
    lavender: '#e6e6fa',
    lavenderblush: '#fff0f5',
    lawngreen: '#7cfc00',
    lemonchiffon: '#fffacd',
    lightblue: '#add8e6',
    lightcoral: '#f08080',
    lightcyan: '#e0ffff',
    lightgoldenrodyellow: '#fafad2',
    lightgray: '#d3d3d3',
    lightgrey: '#d3d3d3',
    lightgreen: '#90ee90',
    lightpink: '#ffb6c1',
    lightsalmon: '#ffa07a',
    lightseagreen: '#20b2aa',
    lightskyblue: '#87cefa',
    lightslategray: '#778899',
    lightslategrey: '#778899',
    lightsteelblue: '#b0c4de',
    lightyellow: '#ffffe0',
    lime: '#00ff00',
    limegreen: '#32cd32',
    linen: '#faf0e6',
    magenta: '#ff00ff',
    maroon: '#800000',
    mediumaquamarine: '#66cdaa',
    mediumblue: '#0000cd',
    mediumorchid: '#ba55d3',
    mediumpurple: '#9370db',
    mediumseagreen: '#3cb371',
    mediumslateblue: '#7b68ee',
    mediumspringgreen: '#00fa9a',
    mediumturquoise: '#48d1cc',
    mediumvioletred: '#c71585',
    midnightblue: '#191970',
    mintcream: '#f5fffa',
    mistyrose: '#ffe4e1',
    moccasin: '#ffe4b5',
    navajowhite: '#ffdead',
    navy: '#000080',
    oldlace: '#fdf5e6',
    olive: '#808000',
    olivedrab: '#6b8e23',
    orange: '#ffa500',
    orangered: '#ff4500',
    orchid: '#da70d6',
    palegoldenrod: '#eee8aa',
    palegreen: '#98fb98',
    paleturquoise: '#afeeee',
    palevioletred: '#db7093',
    papayawhip: '#ffefd5',
    peachpuff: '#ffdab9',
    peru: '#cd853f',
    pink: '#ffc0cb',
    plum: '#dda0dd',
    powderblue: '#b0e0e6',
    purple: '#800080',
    rebeccapurple: '#663399',
    red: '#ff0000',
    rosybrown: '#bc8f8f',
    royalblue: '#4169e1',
    saddlebrown: '#8b4513',
    salmon: '#fa8072',
    sandybrown: '#f4a460',
    seagreen: '#2e8b57',
    seashell: '#fff5ee',
    sienna: '#a0522d',
    silver: '#c0c0c0',
    skyblue: '#87ceeb',
    slateblue: '#6a5acd',
    slategray: '#708090',
    slategrey: '#708090',
    snow: '#fffafa',
    springgreen: '#00ff7f',
    steelblue: '#4682b4',
    tan: '#d2b48c',
    teal: '#008080',
    thistle: '#d8bfd8',
    tomato: '#ff6347',
    turquoise: '#40e0d0',
    violet: '#ee82ee',
    wheat: '#f5deb3',
    white: '#ffffff',
    whitesmoke: '#f5f5f5',
    yellow: '#ffff00',
    yellowgreen: '#9acd32',
};

/**
 * CSS properties that contain color values and should be standardized.
 * Extracted from the W3C CSS specification.
 */
const COLOR_PROPERTIES = new Set([
    'backgroundColor',
    'borderTopColor',
    'borderRightColor',
    'borderBottomColor',
    'borderLeftColor',
    'color',
    'outlineColor',
]);

/**
 * Regular expression for matching hex color triplets (short form).
 * Matches patterns like #RGB and expands to #RRGGBB.
 */
const HEX_TRIPLET_RE = /#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])/;

/**
 * Regular expression for validating hex color format.
 * Matches #RGB or #RRGGBB formats.
 */
const VALID_HEX_COLOR_RE = /^#(?:[0-9a-f]{3}){1,2}$/i;

/**
 * Regular expression for parsing rgba() color strings.
 * Matches: rgba(r, g, b, a) or (r, g, b, a) format
 * where r, g, b are integers [0-255] and a is float [0-1].
 */
const RGBA_COLOR_RE = /^(?:rgba)?\((\d{1,3}),\s?(\d{1,3}),\s?(\d{1,3}),\s?(0|1|0\.\d*)\)$/i;

/**
 * Regular expression for parsing rgb() color strings.
 * Matches: rgb(r, g, b) or (r, g, b) format
 * where r, g, b are integers [0-255].
 */
const RGB_COLOR_RE = /^(?:rgb)?\((0|[1-9]\d{0,2}),\s?(0|[1-9]\d{0,2}),\s?(0|[1-9]\d{0,2})\)$/i;

/**
 * Attempts to parse a string as an rgba color.
 * Expects format: 'rgba(r, g, b, a)' or '(r, g, b, a)'
 * where r, g, b are integers in [0, 255] and a is a float in [0, 1].
 *
 * @param str String to parse
 * @returns RGBA tuple [r, g, b, a] or null if invalid
 */
function maybeParseRgbaColor(str: string): RgbaColor | null {
    const match = str.match(RGBA_COLOR_RE);
    if (!match) {
        return null;
    }

    const r = Number(match[1]);
    const g = Number(match[2]);
    const b = Number(match[3]);
    const a = Number(match[4]);

    if (r >= 0 && r <= 255 && g >= 0 && g <= 255 && b >= 0 && b <= 255 && a >= 0 && a <= 1) {
        return [r, g, b, a];
    }

    return null;
}

/**
 * Attempts to parse a string as an rgb color.
 * Expects format: 'rgb(r, g, b)' or '(r, g, b)'
 * where r, g, b are integers in [0, 255].
 *
 * @param str String to parse
 * @returns RGBA tuple [r, g, b, 1] or null if invalid
 */
function maybeParseRgbColor(str: string): RgbaColor | null {
    const match = str.match(RGB_COLOR_RE);
    if (!match) {
        return null;
    }

    const r = Number(match[1]);
    const g = Number(match[2]);
    const b = Number(match[3]);

    if (r >= 0 && r <= 255 && g >= 0 && g <= 255 && b >= 0 && b <= 255) {
        return [r, g, b, 1];
    }

    return null;
}

/**
 * Converts a hex color or CSS color name to RGBA format.
 * Handles both short (#RGB) and long (#RRGGBB) hex formats,
 * as well as standard CSS color names.
 *
 * @param hexOrColorName Hex color (with or without #) or CSS color name
 * @returns RGBA tuple [r, g, b, a] or null if invalid
 */
function maybeConvertHexOrColorName(hexOrColorName: string): RgbaColor | null {
    const normalized = hexOrColorName.toLowerCase();

    // Try to look up as a color name
    let hex = COLOR_NAMES[normalized];

    if (!hex) {
        // Treat as hex color
        hex = normalized.startsWith('#') ? normalized : `#${normalized}`;

        // Expand short form (#RGB -> #RRGGBB)
        if (hex.length === 4) {
            hex = hex.replace(HEX_TRIPLET_RE, '#$1$1$2$2$3$3');
        }

        // Validate hex format
        if (!VALID_HEX_COLOR_RE.test(hex)) {
            return null;
        }
    }

    // Parse hex to RGB
    const r = parseInt(hex.substring(1, 3), 16);
    const g = parseInt(hex.substring(3, 5), 16);
    const b = parseInt(hex.substring(5, 7), 16);

    return [r, g, b, 1];
}

/**
 * Standardizes a CSS color property value to rgba format.
 * Converts hex, rgb, rgba, and named colors to a consistent rgba() format.
 * If the property is not a color property or the value cannot be parsed,
 * returns the original value unchanged.
 *
 * @param propertyName CSS property name in camelCase
 * @param propertyValue The CSS property value
 * @returns Standardized color value in rgba format, or original value if not a color
 *
 * @example
 * standardizeColor('color', '#f00')           // 'rgba(255, 0, 0, 1)'
 * standardizeColor('color', 'rgb(255, 0, 0)') // 'rgba(255, 0, 0, 1)'
 * standardizeColor('color', 'red')            // 'rgba(255, 0, 0, 1)'
 * standardizeColor('margin', '10px')          // '10px' (not a color property)
 */
export function standardizeColor(propertyName: string, propertyValue: string): string {
    // Only process known color properties
    if (!COLOR_PROPERTIES.has(propertyName)) {
        return propertyValue;
    }

    // Try parsing in order of specificity
    const rgba =
        maybeParseRgbaColor(propertyValue) ||
        maybeParseRgbColor(propertyValue) ||
        maybeConvertHexOrColorName(propertyValue);

    // If parsing succeeded, return standardized rgba format
    if (rgba) {
        return `rgba(${rgba.join(', ')})`;
    }

    // If parsing failed, return original value
    return propertyValue;
}
